samples/tutorial/Tutorial3.md

```diff
@@ -0,0 +1,415 @@
+    // The `Todo` from our parent changed. Update our internal copy so we are starting from the same
+    // item. The "correct" behavior depends on the business logic - would we only want to update if
+    // the users hasn't changed the todo from the initial one? Or is it ok to delete whatever edits
+    // were in progress if the state from the parent changes?
+    if (old.initialTodo != new.initialTodo) {
+      return state.copy(todo = new.initialTodo)
+    }
+    return state
+  }
+
+  // …
+}
+```
+
+Next, define the actions this workflow will handle - specifically the title and note changing from the UI:
+
+```kotlin
+object TodoEditWorkflow : StatefulWorkflow<EditProps, State, Output, TodoEditScreen>() {
+
+  // …
+
+  private fun onTitleChanged(title: String) = action {
+    state = state.withTitle(title)
+  }
+
+  private fun onNoteChanged(note: String) = action {
+    state = state.withNote(note)
+  }
+
+  private fun onDiscard() = action {
+    // Emit the Discard output when the discard action is received.
+    setOutput(Discard)
+  }
+
+  private fun onSave() = action {
+    // Emit the Save output with the current todo state when the save action is received.
+    setOutput(Save(state.todo))
+  }
+
+  private fun State.withTitle(title: String) = copy(todo = todo.copy(title = title))
+  private fun State.withNote(note: String) = copy(todo = todo.copy(note = note))
+```
+
+Finally, update the `render` method to return a `TodoEditScreen`:
+
+```kotlin
+object TodoEditWorkflow : StatefulWorkflow<EditProps, State, Output, TodoEditScreen>() {
+
+  // …
+
+  override fun render(
+    props: EditProps,
+    state: State,
+    context: RenderContext
+  ): TodoEditScreen {
+    return TodoEditScreen(
+        title = state.todo.title,
+        note = state.todo.note,
+        onTitleChanged = { context.actionSink.send(onTitleChanged(it)) },
+        onNoteChanged = { context.actionSink.send(onNoteChanged(it)) },
+        saveChanges = { context.actionSink.send(onSave()) },
+        discardChanges = { context.actionSink.send(onDiscard()) }
+    )
+  }
+
+  // …
+}
+```
+
+Now the workflow provides a backing for the UI to edit a todo item, but it doesn't support saving and discarding changes. Add two `Output`s and actions for these cases:
+
+```kotlin
+object TodoEditWorkflow : StatefulWorkflow<EditProps, State, Output, TodoEditScreen>() {
+
+  // …
+
+  sealed class Output {
+    object Discard : Output()
+    data class Save(val todo: TodoModel) : Output()
+  }
+
+  // …
+}
+```
+
+## Todo Editing in the full flow
+
+### Updating the current workflows to prepare to add the edit workflow
+
+We want the todo edit screen to be shown when a user taps on an item on the todo list screen. To do this, we will modify the todo list workflow to show the edit screen when we are editing.
+
+Because the `TodoEditWorkflow` returns a different screen type, we will first need to modify the todo list workflow to return a list of `Any`s as the rendering.
```

Comment:
It'd be more idiomatic for it to emit `BackStackScreen<Any>`. They have appropriate `plus` operators. Maybe you make that change later?

The sentence is also a little strange -- how does returning a different screen type make us decide to render a `List<Any>` rather than just `Any`?

Reply:
I actually changed it back intentionally. Seems better to me to use more simple, more general types for these children workflows that don't actually need to use the special view binding.

Reply:
> The sentence is also a little strange -- how does returning a different screen type make us decide to render a List<Any> rather than just Any?

That's a good point though. How about: "Because the parent workflow needs to know about the fact that this workflow renders a list, so it can append its screens to the backstack, we will first need to modify the todo list workflow to return a _list_ of `Any`s as the rendering."?

Reply:
Def. better. But still leaves implicit _why_ we're returning a list, I think. Did you say anything like _we're going to return one or two screens that the parent can use to build a back stack_.

Reply:
> The `TodoListWorkflow` will now occasionally need to render two screens instead of just the one. Its parent workflow will add the one or two screens to the backstack it constructs. Since `TodoListWorkflow`'s rendering type is `Any`, it can just return a list of screens. However, the parent workflow needs to know about the list, so it can pull the screens out and add them to its backstack. We'll change the rendering type from `Any` to `List<Any>`:

How's this?